### PR TITLE
Log snmpv3 auth failures

### LINKED ISSUE #7929: Random integers in SNMP Input Plugin OID output
- Telegraf version - 1.15.1
- Operating System - Redhat Linux
- Input Plugin - SNMP 

I'm using the snmp input plugin to poll certain OIDs. 
Every time I run telegraf, I'm getting sequential integers in outputs of OID for random hosts as can be seen below : 

```
viptela_mem_cpu_stats,agent_host=10.14.*.*,hostname=903 `memory_free=909i,uptime=904i,cpu_util_system=905i,cpu_util_user=906i,memory_total=907i,memory_used=908i` 1596221521000000000

viptela_mem_cpu_stats,agent_host=10.12.*.*,hostname=TEST-VE1 memory_used="3502212",memory_free="1761120",uptime=21752169i,cpu_util_system="3.49",cpu_util_user="2.28",memory_total="6081060" 1596221521000000000

viptela_mem_cpu_stats,agent_host=10.13.*.*,hostname=924 uptime=925i,cpu_util_system=926i,cpu_util_user=927i,memory_total=928i,memory_used=929i,memory_free=930i 1596221521000000000
```

### PR DESCRIPTION
**Pull request by Hipska**
While I still [believe](https://github.com/gosnmp/gosnmp/issues/307) this is a check that `gosnmp` should do, let's not wait for them to implement.

Fixes #7929 as no metrics with strange values will be produced, instead telegraf will return a clear error message.
Closes #8215 since there is no feedback from original author.

### Required for all PRs:

- [ ] Associated README.md updated.
- [ ] Has appropriate unit tests.

### CODE REVIEW

**Review by Jagularr — APPROVED** [review]

File: plugins/inputs/snmp/snmp.go
Comment (reimda) [inline plugins/inputs/snmp/snmp.go:420]: I'm new to SNMP v3 report PDUs.  As I understand it there are only a few reasons telegraf would receive a report PDU and one of the reasons is an auth error.  An auth error is very likely due to the user misconfiguring a password.

I think the error message here should suggest the user check their SNMP auth config.  I like how #8915 did this with "received SNMP report instead of response for field %s: check authentication?"

**Review by reimda — CHANGES_REQUESTED** [review]

File: plugins/inputs/snmp/snmp.go
Comment (Hipska) [reply plugins/inputs/snmp/snmp.go:420]: The only reason for my PR instead of the original one was to be able to give a more specific message to the user. Just saying a report received will not help much for a regular user. This is also the reason I try to translate the report OID so I could mention what exactly the report was about. 

In my opinion, If a user sees one of these names in the log message, they might understand what the problem is.
- usmStatsUnsupportedSecLevels
- usmStatsNotInTimeWindows
- usmStatsUnknownUserNames
- usmStatsUnknownEngineIDs
- usmStatsWrongDigests
- usmStatsDecryptionErrors

(some of them, the user will never see since gosnmp is already handling them)

And these are only the ones that occur when having a bad auth setup, I think there are more possibilities when a report could be received. (If not now, it might be in the future or depending on vendor implementation?)

So if you still feel the message on the original PR was better, feel free to merge that one and close this one! I just want 'a' solution to be included ASAP.

File: plugins/inputs/snmp/snmp.go
Comment (Hipska) [reply plugins/inputs/snmp/snmp.go:419]: Oh, okay I wrongly assumed that the original OID would still be returned in case of error. (Just like when the OID could not be translated, because it is unknown) Shouldn't that be more useful?

**Comment by Hipska** [comment]
Hi @reimda I have added some hardcoded messages for the most common types of report that will be received. What do you think of that?

File: plugins/inputs/snmp/snmp.go
Comment (Hipska) [reply plugins/inputs/snmp/snmp.go]: Discussed during call that not checking the `err` isn't a usual go pattern, so I will change that.

**Comment by Hipska** [comment]
There is something going on with https://github.com/gosnmp/gosnmp/pull/314 but I'm not yet convinced it is good yet.

**Comment by Hipska** [comment]
Okay, we can go. When a new version of gosnmp is released, I will update our check..

**Comment by reimda** [comment]
With https://github.com/gosnmp/gosnmp/pull/314 merged it sounds like we should change our approach for this PR.  If we update gosnmp we won't need to hard code the USM report OIDs, and we can use the improved error message that Get() returns instead of adding our own.

**Comment by Hipska** [comment]
Indeed, so maybe we can already merge this and adapt the improvements when gosnmp releases a new version afterwards?

**Comment by reimda** [comment]
I was thinking since the gosnmp PR has been merged and since go mod lets us depend on any commit in a library, we don't need to wait for a new gosnmp release.  We could `go get github.com/gosnmp/gosnmp@a3738642b2d1` to use the latest from gosnmp.  That would let us fix the auth failures now and in the way we expect to use long-term.

I ordinarily avoid using prerelease libraries, but looking at gosnmp's history, besides PR 314 there is only one other code change since the 1.30.0 release and it's pretty simple.  In this case using a prerelease doesn't introduce much risk.

Gosnmp seems to be making releases every few months and 1.30.0 came out a month ago.  If we don't use a prerelease version we will likely have to wait a month or two for a release before we come back to this.

**Comment by TimRots** [comment]
> Gosnmp seems to be making releases every few months and 1.30.0 came out a month ago. If we don't use a prerelease version we will likely have to wait a month or two for a release before we come back to this.

Sorry for the hold-up. Released gosnmp [v1.31.0](https://github.com/gosnmp/gosnmp/releases/tag/v1.31.0) today.

Thanks again @Hipska for your help

**Comment by Hipska** [comment]
Great! I think I’ll have time this week to implement this.

**Comment by reimda** [comment]
> > Gosnmp seems to be making releases every few months and 1.30.0 came out a month ago. If we don't use a prerelease version we will likely have to wait a month or two for a release before we come back to this.
> 
> Sorry for the hold-up. Released gosnmp [v1.31.0](https://github.com/gosnmp/gosnmp/releases/tag/v1.31.0) today.

@TimRots Thanks for your work on gosnmp and for releasing 1.31.0!  I was clearly too pessimistic about release scheduling.  Having a gosnmp release that includes gosnmp/gosnmp#314 is great for this PR.

**Comment by TimRots** [comment]
> @TimRots Thanks for your work on gosnmp and for releasing 1.31.0!

Thank you @reimda. Happy to help out :)

>  I was clearly too pessimistic about release scheduling.

Since gosnmp became a community effort we try to do a monthly release, but it still depends on how much time we have to work on things. If a bug is urgent and we try to move forward quicker ofc.

> Having a gosnmp release that includes [gosnmp/gosnmp#314](https://github.com/gosnmp/gosnmp/pull/314) is great for this PR.

314 landed really nice with the help from hipska and superq. Will keep an eye on this pr to see how it works out here :)

**Comment by Hipska** [comment]
Hi @reimda, can you please review again?

**Review by reimda — APPROVED** [review]
Looks great!

**Review by srebhan — APPROVED** [review]
Looks good to me.

**Comment by Stomper98** [comment]
I am still seeing this issue on Centos 7 with Telegraf 1.21.2. We get massive amount of hostnames with random numeric values when a remote SNMPv3 service fails connection. Following through on all the tickets about this issue with gosnmp, is there a fix in sight ?

**Comment by Hipska** [comment]
@Stomper98 whoosh this is a fix/Pr that has been merged a very long time ago. So on our end there is nothing to be fixed since we didn’t got any reports on it anymore.

if you experience any snmpv3 issues, please create a new bug with your relevant config and log messages, so we can look into it.

**Pull request merged**
